<a href="https://colab.research.google.com/github/redadmiral/python-for-journalists/blob/main/exercises/solutions/UebungScraperSelenium.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Übung: Scraping mit Selenium

1. Versuche die Seite https://www.amazon.de mit Python `requests` herunterzuladen. Lass dir den Status Code des Requests ausgeben. Wie interpretierst du den Status Code?



In [1]:
import requests

requests.get("https://www.amazon.de")

<Response [503]>

Status Code 503 steht für Service Anavailable. Nachdem die Seite über den Browser funktioniert lässt Amazon automatisierte Anfragen nicht zu.

2. Installiere die nötigen Abhängigkeiten um Selenium zu benutzen. Orientiere dich dafür an der Anleitung im Video.

2.1 Installiere selenium mit `pip` und lade dir die aktuelle Version von Chrome herunter. Installiere das Paket mit `apt`.

In [ ]:
!pip install selenium
!wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
!apt install ./google-chrome-stable_current_amd64.deb

2.2 Überprüfe die aktuelle Version von Google Chrome.

In [3]:
!google-chrome --version

Google Chrome 113.0.5672.126 


2.3 Installiere die korrekte Version des Chromedrivers.

In [4]:
!wget https://chromedriver.storage.googleapis.com/110.0.5481.77/chromedriver_linux64.zip
!unzip chromedriver_linux64.zip

--2023-05-29 18:48:00--  https://chromedriver.storage.googleapis.com/110.0.5481.77/chromedriver_linux64.zip
Resolving chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)... 74.125.204.128, 2404:6800:4008:c04::80
Connecting to chromedriver.storage.googleapis.com (chromedriver.storage.googleapis.com)|74.125.204.128|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 7396711 (7.1M) [application/zip]
Saving to: ‘chromedriver_linux64.zip’

chromedriver_linux6 100%[===================>]   7.05M  6.91MB/s    in 1.0s    

2023-05-29 18:48:02 (6.91 MB/s) - ‘chromedriver_linux64.zip’ saved [7396711/7396711]

Archive:  chromedriver_linux64.zip
  inflating: chromedriver            
  inflating: LICENSE.chromedriver    


2.4 Erstelle eine Variable `browser` in der du eine Instanz von Chrome ablegst. Übergib ihm die Optionen `--headless` und `--no-sandbox`, wie im Video gezeigt.

In [65]:
from selenium import webdriver

options = webdriver.ChromeOptions()
options.add_argument('--headless')
options.add_argument('--no-sandbox')

browser = webdriver.Chrome(executable_path='chromedriver',options=options)

3.1 Rufe die Seite https://www.amazon.de mit Selenium auf. Nutze die `get`-Methode des Browsers.

In [66]:
browser.get("https://www.amazon.de")

In [67]:
assert browser.title.startswith("Amazon")

3.2 Speichere einen Screenshot der Startseite von amazon.de unter dem namen `amazon_screenshot.png`.

In [68]:
browser.save_screenshot("amazon_screenshot.png")

True

In [69]:
import os

assert os.path.isfile("amazon_screenshot.png")
assert os.path.getsize("amazon_screenshot.png") > 1000

Exkurs: Manchmal poppt hier das Cookie-Bannera auf. Wähle in dem Fall den Accept Button mit der ID `sp-cc-accept` aus. Speichere ihn in der Variable `accept` und wähle die `click()`-Methode aus um das Banner zu bestätigen. 

In [70]:
from selenium.webdriver.common.by import By

accept = browser.find_element(By.ID, "sp-cc-accept")
accept.click()

In [71]:
browser.save_screenshot("amazon_screenshot_after_cookie.png")

True

3.3 Wähle das Suchfeld aus und speichere es in der Variable `search_bar`.

In [72]:
from selenium.webdriver.common.by import By

search_bar = browser.find_element(By.ID, "twotabsearchtextbox")

In [73]:
assert search_bar.text == ""

3.4 Trage das Wort "Hundefutter" in die Suchleiste ein.

In [74]:
search_bar.send_keys("Hundefutter")

In [75]:
browser.save_screenshot("search_bar_filled.png")

True

3.5 Führe die Suche aus. Das kannst du auf zwei Weisen machen:

1. Wähle das Suchfeld mit der Lupe aus und führe die `click()`-Methode aus.
2. Importiere 

In [76]:
search_button = browser.find_element(By.ID, "nav-search-submit-button")
search_button.click()

In [77]:
browser.save_screenshot("search_result.png")

True

Falls die Seite hier eine Fehlermeldung anzeigt ist das nicht weiter schlimm.